### PR TITLE
FGAT and 4DEnVar related changes for FV3 regional DA

### PR DESCRIPTION
**Pull request by BinLiu-NOAA**
From Xu Lu and Xuguang Wang (OU): FGAT and 4DEnVar related changes for FV3 regional DA.

### CODE REVIEW

File: src/gsi/fv3_regional_interface.f90 [hunk 1]
```diff
@@ -18,7 +18,7 @@ subroutine convert_fv3_regional
 !   2017-04-08  parrish
 !   2018-02-16  wu   - read in grid and time infor from fv3 files
 !                      read directly from fv3 files and not writeout GSI internal file 
-!
+!   2020-11-19  Xu Lu & Xuguang Wang - add time label it for fgat. POC: xuguang.wang@ou.edu
```
Comment (lbi2018) [inline src/gsi/fv3_regional_interface.f90]: Please only leave last name in the document line

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 2]
```diff
@@ -12,6 +12,7 @@ module gsi_rfv3io_mod
 !   2017-10-10  wu      - setup A grid and interpolation coeff in generate_anl_grid
 !   2018-02-22  wu      - add subroutines for read/write fv3_ncdf
 !   2019        ting    - modifications for use for ensemble IO and cold start files 
+!   2020-11-19 Xu Lu & Xuguang Wang - add time label it for fgat/4denvar. POC: xuguang.wang@ou.edu
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Please only leave last name in the document session

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 3]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
 
 contains
-  subroutine fv3regfilename_init(this,grid_spec_input,ak_bk_input,dynvars_input, &
+  subroutine fv3regfilename_init(this,it,grid_spec_input,ak_bk_input,dynvars_input, &
                       tracers_input,sfcdata_input,couplerres_input)
   implicit None
   class(type_fv3regfilenameg),intent(inout):: this
   character(*),optional :: grid_spec_input,ak_bk_input,dynvars_input, &
                       tracers_input,sfcdata_input,couplerres_input
+  integer,intent(in) :: it ! OU default +-3 hours for FGAT
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Please remove document after the "!"

File: src/gsi/read_guess.F90 [hunk 4]
```diff
@@ -75,6 +75,7 @@ subroutine read_guess(iyear,month,idd,mype)
 !   2017-10-10  Wu W    - add code for FV3 netcdf guess input 
 !   2019-09-18  martin  - added new fields to save guess tsen, q, geop_hgt for writing increment
 !   2019-09-23  martin  - add code for FV3 GFS netcdf guess input
+!   2020-11-19  Xu Lu & Xuguang Wang - add time label it_c for fgat/4denvar read in. POC: xuguang.wang@ou.edu
```
Comment (lbi2018) [inline src/gsi/read_guess.F90]: Please only leave last name in the document session

File: src/gsi/read_guess.F90 [hunk 5]
```diff
@@ -164,8 +167,28 @@ subroutine read_guess(iyear,month,idd,mype)
         else if (nems_nmmb_regional) then
            call nmm_binary_guess%read_nems_nmmb_guess(mype)
         else if (fv3_regional      ) then
-           call bg_fv3regfilenameg%init
-           call  read_fv3_netcdf_guess(bg_fv3regfilenameg)
+           it_c=0
+           do it=1,7 !Xu default only -+3 hours for FGAT
```
Comment (lbi2018) [inline src/gsi/read_guess.F90]: Please remove the document after "!"

File: src/gsi/read_guess.F90 [hunk 6]
```diff
@@ -164,8 +167,28 @@ subroutine read_guess(iyear,month,idd,mype)
...
+                 if(.not.file_exists) then
+                    if (mype.eq.0) write(6,*)' file ',trim(filename),' not exists for FGAT!'
+                    cycle
+                 endif
+              endif
+              it_c=it_c+1
+              if (mype.eq.0) print *,filename,"file read in for first guess",nfldsig
+              call bg_fv3regfilenameg%init(it)
+              call read_fv3_netcdf_guess(bg_fv3regfilenameg,it_c)
+           end do !Xu
```
Comment (lbi2018) [inline src/gsi/read_guess.F90]: Please remove !Xu in line 191

File: src/gsi/write_all.F90 [hunk 7]
```diff
@@ -96,6 +98,7 @@ subroutine write_all(increment)
 !   2013-10-19  todling - metguess holds ges fields now
 !   2014-10-05  todling - background biases now held in bundle
 !   2017-10-10  Wu W    - add FV3 option for regional output
+!   2020-11-19  Xu Lu & Xuguang Wang - add option to write 4denvar analyses at all times. POC: xuguang.wang@ou.edu
```
Comment (lbi2018) [inline src/gsi/write_all.F90]: Please only leave last name here.

File: src/gsi/write_all.F90 [hunk 8]
```diff
@@ -120,7 +125,26 @@ subroutine write_all(increment)
 ! Regional output
   if (regional) then
      if (fv3_regional) then
-        call wrfv3_netcdf(bg_fv3regfilenameg)
+        it_c=0
+        do it=1,7 !OU default only -+3 hours for FGAT
```
Comment (lbi2018) [inline src/gsi/write_all.F90]: Please remove the document after "!"

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 9]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
+  character(255):: filename
   if(present(grid_spec_input))then
 
     this%grid_spec=grid_spec_input
   else
+   if (it.eq.4) then
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Combine line 115 and line 116

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 10]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
     this%grid_spec='fv3_grid_spec'
+   else
+    write(filename,"(A14,I2.2)") 'fv3_grid_spec_',(it+2)
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Add indentation here.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 11]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
+    this%grid_spec=trim(filename)
+   end if
   endif
   if(present(ak_bk_input))then
     this%ak_bk=ak_bk_input
   else
+   if (it.eq.4) then
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Combine line 125 and 126.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 12]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
     this%ak_bk='fv3_ak_bk'
+   else
+    write(filename,"(A10,I2.2)") 'fv3_ak_bk_',(it+2)
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Add indentation here.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 13]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
+    this%ak_bk=trim(filename)
+   endif
   endif
   if(present(dynvars_input))then
 
     this%dynvars=dynvars_input
   else
+   if (it.eq.4) then
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Combine line 136 and 137.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 14]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
+   else
+    write(filename,"(A12,I2.2)") 'fv3_dynvars_',(it+2)
+    this%dynvars=trim(filename)
+   endif
   endif
   if(present(tracers_input))then
 
     this%tracers=tracers_input
   else
+   if (it.eq.4) then
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Combine line 147 and 148.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 15]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
+   endif
   endif
   if(present(sfcdata_input))then
 
     this%sfcdata=sfcdata_input
   else
+   if (it.eq.4) then
     this%sfcdata='fv3_sfcdata'
+   else
+    write(filename,"(A12,I2.2)") 'fv3_sfcdata_',(it+2)
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Add more indentation here.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 16]
```diff
@@ -100,47 +101,79 @@ module gsi_rfv3io_mod
...
   endif
 
   if(present(couplerres_input))then
 
     this%couplerres=couplerres_input
   else
+   if (it.eq.4) then
     this%couplerres='coupler.res'
+   else
+    write(filename,"(A12,I2.2)") 'coupler.res_',(it+2)
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Add more indentation here.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 17]
```diff
@@ -528,14 +562,21 @@ subroutine read_fv3_files(mype)
        in_unit=15
        iwan=0
 !WWWWWW setup for one first guess file for now
-!      do i=0,9 !place holder for FGAT
-       i=3
+       do i=0,99 !place holder for FGAT
+!OU       i=3
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Remove line 566.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 18]
```diff
@@ -528,14 +562,21 @@ subroutine read_fv3_files(mype)
...
+          if (i.eq.6) then
+           write(filenames,"(A11)") 'fv3_dynvars'
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Add more indentation here.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 19]
```diff
@@ -528,14 +562,21 @@ subroutine read_fv3_files(mype)
...
+          endif
+          INQUIRE(FILE=filenames, EXIST=fexist)
+          if(.not.fexist) cycle
 
 !wwww read in from the external file directly, no internal files sigfxx for FV3
           idate5(1)=  regional_time(1)
           idate5(2)=  regional_time(2)
           idate5(3)=  regional_time(3)
-          idate5(4)=  regional_time(4)
+          idate5(4)=  regional_time(4) + i - 6
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Not sure if I understand line 579.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 20]
```diff
@@ -583,12 +624,18 @@ subroutine read_fv3_files(mype)
 !    Check for consistency of times from surface guess files.
        iwan=0
        do i=0,99
-          write(filename,200)i
-  200     format('sfcf',i2.2)
-          inquire(file=filename,exist=fexist)
+!         write(filename,200)i
+! 200     format('sfcf',i2.2)
+!         inquire(file=filename,exist=fexist)
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Line 627-629 have been removed?

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 21]
```diff
@@ -583,12 +624,18 @@ subroutine read_fv3_files(mype)
...
+          if(i.eq.6) then
+             write(filenames,"(A11)") 'fv3_sfcdata'
+          else
+             write(filenames,"(A12,I2.2)") 'fv3_sfcdata_',i
+          endif
+          INQUIRE(FILE=filenames, EXIST=fexist)
           if(fexist)then
              idateg(4)=iadate(1); idateg(2)=iadate(2)
-             idateg(3)=iadate(3); idateg(1)=iadate(4)
+             idateg(3)=iadate(3); idateg(1)=iadate(4) + i - 6
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Not sure if I understand line 638.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 22]
```diff
@@ -749,8 +797,14 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
     mype_oz=6
     mype_2d=7 
       
-    allocate(ijns(npe),ijns2d(npe),ijnz(npe) )
-    allocate(displss(npe),displss2d(npe),displsz_g(npe) )
+!   allocate(ijns(npe),ijns2d(npe),ijnz(npe) )
+!   allocate(displss(npe),displss2d(npe),displsz_g(npe) )
+    if (.not.allocated(ijns)) allocate(ijns(npe))
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90:797]: Line 802 to 807, are these code changes for FGAT?

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 23]
```diff
@@ -793,16 +847,17 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
     endif
 
     if( fv3sar_bg_opt == 0) then 
-       call gsi_fv3ncdf_read(dynvars,'DELP','delp',ges_prsi,mype_p)
+!      call gsi_fv3ncdf_read(dynvars,'DELP','delp',ges_prsi,mype_p)
+       call gsi_fv3ncdf_read(dynvars,'DELP','delp',ges_prsi(1,1,1,it),mype_p)
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: Have concern for line 851.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 24]
```diff
@@ -793,16 +847,17 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
...
        ges_prsi(:,:,nsig+1,it)=eta1_ll(nsig+1)
        do i=nsig,1,-1
           ges_prsi(:,:,i,it)=ges_prsi(:,:,i,it)*0.001_r_kind+ges_prsi(:,:,i+1,it)
        enddo
        ges_ps(:,:)=ges_prsi(:,:,1,it)
     else  
        call  gsi_fv3ncdf2d_read_v1(dynvars,'ps','PS',ges_ps,mype_p)
-       ges_ps=ges_ps*0.001_r_kind
        ges_prsi(:,:,nsig+1,it)=eta1_ll(nsig+1)
+       ges_ps=ges_ps*0.001_r_kind
```
Comment (lbi2018) [inline src/gsi/gsi_rfv3io_mod.f90]: What's the reason to change order of line 859 and 860?

File: src/gsi/read_guess.F90 (hunk 6 above)
Comment (lbi2018) [inline src/gsi/read_guess.F90]: Please remove !xu

File: src/gsi/write_all.F90 [hunk 25]
```diff
@@ -120,7 +125,26 @@ subroutine write_all(increment)
...
+           else
+              write(filename,"(A12,I2.2)") 'fv3_sfcdata_',(it+2)
+              INQUIRE(FILE=filename, EXIST=file_exists)
+              if(.not.file_exists) then
+                 cycle
+              endif
+              if (.not.lwrite4danl) cycle
+           endif
+           it_c=it_c+1
+           call bg_fv3regfilenameg%init(it) !OU default 4 for +- 3 hours FGAT
```
Comment (lbi2018) [inline src/gsi/write_all.F90]: Remove document after ! for line 145

**Review by lbi2018 — CHANGES_REQUESTED** [review]

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 19 above)
Comment (TingLei-NOAA) [inline src/gsi/gsi_rfv3io_mod.f90]: This line is supposed to give the hours . However, this item is supposed to be the hour in a day.  The current line may give the number outside of 0-24 .  In that situation, the code can still work?   I am not sure about this.  Hence, some transformation may be needed. I guess there are some function in w3fs21 lib to help on this.

File: src/gsi/read_guess.F90 [hunk 26]
```diff
@@ -164,8 +167,28 @@ subroutine read_guess(iyear,month,idd,mype)
         else if (nems_nmmb_regional) then
            call nmm_binary_guess%read_nems_nmmb_guess(mype)
         else if (fv3_regional      ) then
-           call bg_fv3regfilenameg%init
-           call  read_fv3_netcdf_guess(bg_fv3regfilenameg)
+           it_c=0
+           do it=1,7 !Xu default only -+3 hours for FGAT
+              if(it.eq.4) then
```
Comment (TingLei-NOAA) [inline src/gsi/read_guess.F90]: I suggest to put the nfldsig loop of reading ( for it) into the read_fv3_netcdfguess subroutine, as does regional wrf does. 
Another problem is about the naming convention of fv3reg files on different time levels.  Can we make it be compatible to the "old runs" or we need a new convention?  like we need to set the name of bg files on the analysis time : fv3_dynvar_01 (for hourly da ,and so on)?

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 22 above)
Comment (XuLu-NOAA) [reply src/gsi/gsi_rfv3io_mod.f90:797]: Yes, the variables will be allocated multiple times if not treated when there are multiple first guesses read in

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 24 above)
Comment (XuLu-NOAA) [reply src/gsi/gsi_rfv3io_mod.f90]: No need to. It must be an error.

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 19 above)
Comment (XuLu-NOAA) [reply src/gsi/gsi_rfv3io_mod.f90]: Yes, good catch.  Modified using the w3fs15.

File: src/gsi/read_guess.F90 (hunk 26 above)
Comment (XuLu-NOAA) [reply src/gsi/read_guess.F90]: Thanks for the suggestion. The read_guess has been modified accordingly using the nfldsig.  
But the bg files was named as fv3_dynvars_03~fv3_dynvars_09, not sure about the last question.

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 19 above)
Comment (XuLu-NOAA) [reply src/gsi/gsi_rfv3io_mod.f90]: As indicated by Ting, this is a bugged modification for time.  Further modified using the w3fs15 function.

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 21 above)
Comment (XuLu-NOAA) [reply src/gsi/gsi_rfv3io_mod.f90]: Same as L579

**Pull request merged**